### PR TITLE
[SPARK-32444][SQL] Infer filters from DPP

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This pr add support infer filters from DPP.

For [this test suite](https://github.com/apache/spark/blob/69be17bfc3845668fd8f5aed8026c04da4496677/sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala#L1285-L1297).
Table `fact_stats` is partitioned by `store_id` and table `code_stats` is partitioned by `store_id`.
DPP add a new predicate: `fact_stats.store_id IN dynamicpruning#2723`.
We can infer `code_stats.store_id IN dynamicpruning#2723` base on [`t1.store_id = t2.store_id`](https://github.com/apache/spark/blob/69be17bfc3845668fd8f5aed8026c04da4496677/sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala#L1294).

### Why are the changes needed?

Improve query performance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by bart-samwel** [comment]
The partition pruning already prunes all partitions that don't satisfy that predicate, so all remaining rows will satisfy that predicate, right? Where is the performance gain then?

**Comment by wangyum** [comment]
Before this pr. We can only prune on `fact_stats.store_id` column:
```
== Physical Plan ==
*(3) Project [store_id#2705, code#2706, product_id#2708L]
+- *(3) BroadcastHashJoin [cast(store_id#2705 as bigint)], [store_id#2709L], Inner, BuildRight, false
   :- *(3) Project [store_id#2705, code#2706]
   :  +- *(3) BroadcastHashJoin [store_id#2705], [store_id#2707], Inner, BuildRight, false
   :     :- *(3) ColumnarToRow
   :     :  +- FileScan parquet default.fact_stats[store_id#2705] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex[file:/Users/yumwang/spark/SPARK-27227/sql/core/spark-warehouse/org.apache.spark..., PartitionFilters: [isnotnull(store_id#2705), dynamicpruningexpression(cast(store_id#2705 as bigint) IN subquery#2716)], PushedFilters: [], ReadSchema: struct<>
   :     :        +- Subquery subquery#2716, [id=#245]
   :     :           +- *(2) HashAggregate(keys=[store_id#2709L#2715L], functions=[])
   :     :              +- Exchange hashpartitioning(store_id#2709L#2715L, 5), true, [id=#241]
   :     :                 +- *(1) HashAggregate(keys=[store_id#2709L AS store_id#2709L#2715L], functions=[])
   :     :                    +- *(1) Filter ((isnotnull(product_id#2708L) AND (product_id#2708L < 3)) AND isnotnull(store_id#2709L))
   :     :                       +- *(1) ColumnarToRow
   :     :                          +- FileScan parquet default.product[product_id#2708L,store_id#2709L] Batched: true, DataFilters: [isnotnull(product_id#2708L), (product_id#2708L < 3), isnotnull(store_id#2709L)], Format: Parquet, Location: InMemoryFileIndex[file:/Users/yumwang/spark/SPARK-27227/sql/core/spark-warehouse/org.apache.spark..., PartitionFilters: [], PushedFilters: [IsNotNull(product_id), LessThan(product_id,3), IsNotNull(store_id)], ReadSchema: struct<product_id:bigint,store_id:bigint>
   :     +- BroadcastExchange HashedRelationBroadcastMode(List(cast(input[1, int, true] as bigint)),false), [id=#275]
   :        +- *(1) ColumnarToRow
   :           +- FileScan parquet default.code_stats[code#2706,store_id#2707] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex[file:/Users/yumwang/spark/SPARK-27227/sql/core/spark-warehouse/org.apache.spark..., PartitionFilters: [isnotnull(store_id#2707)], PushedFilters: [], ReadSchema: struct<code:int>
   +- BroadcastExchange HashedRelationBroadcastMode(List(input[1, bigint, false]),false), [id=#283]
      +- *(2) Filter ((isnotnull(product_id#2708L) AND (product_id#2708L < 3)) AND isnotnull(store_id#2709L))
         +- *(2) ColumnarToRow
            +- FileScan parquet default.product[product_id#2708L,store_id#2709L] Batched: true, DataFilters: [isnotnull(product_id#2708L), (product_id#2708L < 3), isnotnull(store_id#2709L)], Format: Parquet, Location: InMemoryFileIndex[file:/Users/yumwang/spark/SPARK-27227/sql/core/spark-warehouse/org.apache.spark..., PartitionFilters: [], PushedFilters: [IsNotNull(product_id), LessThan(product_id,3), IsNotNull(store_id)], ReadSchema: struct<product_id:bigint,store_id:bigint>
```

After this pr. We can also prune on `code_stats.store_id` column:
```
== Physical Plan ==
*(3) Project [store_id#2705, code#2706, product_id#2708L]
+- *(3) BroadcastHashJoin [cast(store_id#2705 as bigint)], [store_id#2709L], Inner, BuildRight, false
   :- *(3) Project [store_id#2705, code#2706]
   :  +- *(3) BroadcastHashJoin [store_id#2705], [store_id#2707], Inner, BuildRight, false
   :     :- *(3) ColumnarToRow
   :     :  +- FileScan parquet default.fact_stats[store_id#2705] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex[file:/Users/yumwang/spark/SPARK-27227/sql/core/spark-warehouse/org.apache.spark..., PartitionFilters: [isnotnull(store_id#2705), dynamicpruningexpression(cast(store_id#2705 as bigint) IN subquery#2716)], PushedFilters: [], ReadSchema: struct<>
   :     :        +- Subquery subquery#2716, [id=#250]
   :     :           +- *(2) HashAggregate(keys=[store_id#2709L#2715L], functions=[])
   :     :              +- Exchange hashpartitioning(store_id#2709L#2715L, 5), true, [id=#246]
   :     :                 +- *(1) HashAggregate(keys=[store_id#2709L AS store_id#2709L#2715L], functions=[])
   :     :                    +- *(1) Filter ((isnotnull(product_id#2708L) AND (product_id#2708L < 3)) AND isnotnull(store_id#2709L))
   :     :                       +- *(1) ColumnarToRow
   :     :                          +- FileScan parquet default.product[product_id#2708L,store_id#2709L] Batched: true, DataFilters: [isnotnull(product_id#2708L), (product_id#2708L < 3), isnotnull(store_id#2709L)], Format: Parquet, Location: InMemoryFileIndex[file:/Users/yumwang/spark/SPARK-27227/sql/core/spark-warehouse/org.apache.spark..., PartitionFilters: [], PushedFilters: [IsNotNull(product_id), LessThan(product_id,3), IsNotNull(store_id)], ReadSchema: struct<product_id:bigint,store_id:bigint>
   :     +- BroadcastExchange HashedRelationBroadcastMode(List(cast(input[1, int, true] as bigint)),false), [id=#309]
   :        +- *(1) ColumnarToRow
   :           +- FileScan parquet default.code_stats[code#2706,store_id#2707] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex[file:/Users/yumwang/spark/SPARK-27227/sql/core/spark-warehouse/org.apache.spark..., PartitionFilters: [isnotnull(store_id#2707), dynamicpruningexpression(cast(store_id#2707 as bigint) IN subquery#2718)], PushedFilters: [], ReadSchema: struct<code:int>
   :                 +- Subquery subquery#2718, [id=#279]
   :                    +- *(2) HashAggregate(keys=[store_id#2709L#2717L], functions=[])
   :                       +- Exchange hashpartitioning(store_id#2709L#2717L, 5), true, [id=#275]
   :                          +- *(1) HashAggregate(keys=[store_id#2709L AS store_id#2709L#2717L], functions=[])
   :                             +- *(1) Filter ((isnotnull(product_id#2708L) AND (product_id#2708L < 3)) AND isnotnull(store_id#2709L))
   :                                +- *(1) ColumnarToRow
   :                                   +- FileScan parquet default.product[product_id#2708L,store_id#2709L] Batched: true, DataFilters: [isnotnull(product_id#2708L), (product_id#2708L < 3), isnotnull(store_id#2709L)], Format: Parquet, Location: InMemoryFileIndex[file:/Users/yumwang/spark/SPARK-27227/sql/core/spark-warehouse/org.apache.spark..., PartitionFilters: [], PushedFilters: [IsNotNull(product_id), LessThan(product_id,3), IsNotNull(store_id)], ReadSchema: struct<product_id:bigint,store_id:bigint>
   +- BroadcastExchange HashedRelationBroadcastMode(List(input[1, bigint, false]),false), [id=#317]
      +- *(2) Filter ((isnotnull(product_id#2708L) AND (product_id#2708L < 3)) AND isnotnull(store_id#2709L))
         +- *(2) ColumnarToRow
            +- FileScan parquet default.product[product_id#2708L,store_id#2709L] Batched: true, DataFilters: [isnotnull(product_id#2708L), (product_id#2708L < 3), isnotnull(store_id#2709L)], Format: Parquet, Location: InMemoryFileIndex[file:/Users/yumwang/spark/SPARK-27227/sql/core/spark-warehouse/org.apache.spark..., PartitionFilters: [], PushedFilters: [IsNotNull(product_id), LessThan(product_id,3), IsNotNull(store_id)], ReadSchema: struct<product_id:bigint,store_id:bigint>

```

**Comment by bart-samwel** [comment]
Ah, that makes sense. Thanks for the example!

**Comment by maryannxue** [comment]
We don't just apply DPP to any column, right? Shouldn't we check that the inferred filter is applied on partition columns?

In that sense we should fix the DPP rule for it to be apply the same dimension table filter on partition columns from different tables. In other words, the DPP rule itself should generate a filter on `fact_stats.store_id` and another on `code_stats.store_id`.

**Comment by wangyum** [comment]
Yes. We should only infer the filter to partition column.

 The DPP rule itself could not generate a filter on `code_stats.store_id` because it should be after `PushDownPredicates`.

I add logic to `CleanupDynamicPruningFilters`  to remove all filters with `DynamicPruning` that are not filtered on the partition column.

File: sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q16/explain.txt [hunk 1]
```diff
@@ -1,235 +1,247 @@
...
+                                 +- * ColumnarToRow (31)
+                                    +- Scan parquet default.call_center (30)
 
 
 (1) Scan parquet default.catalog_sales
 Output [7]: [cs_ship_date_sk#1, cs_ship_addr_sk#2, cs_call_center_sk#3, cs_warehouse_sk#4, cs_order_number#5, cs_ext_ship_cost#6, cs_net_profit#7]
 Batched: true
 Location: InMemoryFileIndex [file:/Users/yi.wu/IdeaProjects/spark/sql/core/spark-warehouse/org.apache.spark.sql.TPCDSV1_4_PlanStabilitySuite/catalog_sales]
-PushedFilters: [IsNotNull(cs_ship_date_sk), IsNotNull(cs_ship_addr_sk), IsNotNull(cs_call_center_sk)]
+PushedFilters: [IsNotNull(cs_ship_date_sk), IsNotNull(cs_ship_addr_sk), IsNotNull(cs_call_center_sk), IsNotNull(cs_order_number), IsNotNull(cs_warehouse_sk)]
```
Comment (wangyum) [inline sql/core/src/test/resources/tpcds-plan-stability/approved-plans-v1_4/q16/explain.txt:51]: The inferred `IsNotNull` can filter more data.
Before this change:
image



After this change:
image

**Comment by wangyum** [comment]
I have fixed `pruningHasBenefit` for the new filter, and this change will infer more filters for subquery.

**Comment by wangyum** [comment]
This change is similar to #22778.

**Comment by wangyum** [comment]
retest this please

**Pull request closed without merging**